### PR TITLE
Implement promote

### PR DESCRIPTION
**Pull request by jacob-hughes**

### CODE REVIEW

**Comment by jacob-hughes** [comment]
This will require https://github.com/ykjit/ykllvm/pull/185 to land first.

File: ykrt/src/compile/jitc_yk/aot_ir.rs [hunk 1]
```diff
@@ -1322,6 +1329,14 @@ impl fmt::Display for DisplayableInst<'_> {
             Inst::FCmp { lhs, pred, rhs, .. } => {
                 write!(f, "{pred} {}, {}", lhs.display(self.m), rhs.display(self.m))
             }
+            Inst::Promote { val, safepoint, .. } => {
+                write!(
+                    f,
+                    "promote({}) {}",
```
Comment (vext01) [inline ykrt/src/compile/jitc_yk/aot_ir.rs:1335]: I'd say remove the `()` around the promoted operand, so this instruction format like the other AOT instructions?

**Comment by vext01** [comment]
Bar one small comment (and some bikeshedding which I managed to surpress :) ), this looks good Jake!

(I reckon you will need https://github.com/ykjit/yk/pull/1320 to merge before CI will work again -- the corresponding ykllvm change landed, but not the runtime stuff yet)

File: ykrt/src/compile/jitc_yk/trace_builder.rs [hunk 2]
```diff
@@ -62,6 +62,10 @@ pub(crate) struct TraceBuilder {
     // Current count of recursive calls to the function in which outlining was started. Will be 0
     // if `outline_target_blk` is None.
     recursion_count: usize,
+    // Values promoted to constants during runtime. This is only needed during trace building and
+    // can be thrown away after.
+    promotions: Box<[usize]>,
+    promote_idx: usize,
```
Comment (ltratt) [inline ykrt/src/compile/jitc_yk/trace_builder.rs:68]: Can we document this field? [I think I know what it does, but I'll forget in the future!]

Comment (jacob-hughes) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:69]: Sure, fixed here (be9ae27668b8ae05209c1209974434ec35c2dbf0)

File: ykrt/src/compile/jitc_yk/aot_ir.rs (hunk 1 above)
Comment (jacob-hughes) [reply ykrt/src/compile/jitc_yk/aot_ir.rs]: Sure, fixed here (ba681044430dafa8ef109f085623839c77043211)

File: ykrt/src/compile/jitc_yk/trace_builder.rs (hunk 2 above)
Comment (ltratt) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:69]: It's not the "value" is it? I thought it was "what position we are in the promotions array" or similar?

Comment (jacob-hughes) [reply ykrt/src/compile/jitc_yk/trace_builder.rs:69]: How about this? 44851aa4db32ec9a076ec9bbef23ec71f5a9c35b

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by ltratt** [comment]
You'll also need to rebase.

**Comment by jacob-hughes** [comment]
Ok something went a bit weird in that rebase, but I think it's fixed now.

**Comment by jacob-hughes** [comment]
This one can be merged now.

**Comment by ltratt** [comment]
@jacob-hughes The `ptradd.newcg.c` test is failing deterministically:

```
jitstate: start-tracing
i=4, val=3, p=4
jitstate: stop-tracing
thread '<unnamed>' panicked at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/typed-index-collections-3.1.0/src/slice/slice_index.rs:99:10:
index out of bounds: the len is 23 but the index is 2323048404212850720
stack backtrace:
   0: rust_begin_unwind
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/std/src/panicking.rs:665:5
   1: core::panicking::panic_fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/panicking.rs:74:14
   2: core::panicking::panic_bounds_check
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/panicking.rs:276:5
   3: <K as typed_index_collections::slice::slice_index::TiSliceIndex<K,V>>::index
             at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/typed-index-collections-3.1.0/src/slice/slice_index.rs:99:10
   4: <typed_index_collections::slice::TiSlice<K,V> as core::ops::index::Index<I>>::index
             at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/typed-index-collections-3.1.0/src/slice/mod.rs:1507:9
   5: ykrt::compile::jitc_yk::aot_ir::Module::type_
             at /vol/extra/ltratt/yk/ykrt/src/compile/jitc_yk/aot_ir.rs:175:20
   6: ykrt::compile::jitc_yk::aot_ir::Inst::def_type
             at /vol/extra/ltratt/yk/ykrt/src/compile/jitc_yk/aot_ir.rs:1040:26
   7: <ykrt::compile::jitc_yk::aot_ir::DisplayableInst as core::fmt::Display>::fmt
             at /vol/extra/ltratt/yk/ykrt/src/compile/jitc_yk/aot_ir.rs:1141:26
   8: core::fmt::rt::Argument::fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/rt.rs:173:76
   9: core::fmt::write
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:1182:21
  10: core::fmt::Formatter::write_fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:1654:13
  11: <ykrt::compile::jitc_yk::aot_ir::DisplayableBBlock as core::fmt::Display>::fmt
             at /vol/extra/ltratt/yk/ykrt/src/compile/jitc_yk/aot_ir.rs:1394:13
  12: core::fmt::rt::Argument::fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/rt.rs:173:76
  13: core::fmt::write
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:1182:21
  14: core::fmt::Formatter::write_fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:1654:13
  15: <ykrt::compile::jitc_yk::aot_ir::DisplayableFunc as core::fmt::Display>::fmt
             at /vol/extra/ltratt/yk/ykrt/src/compile/jitc_yk/aot_ir.rs:1504:21
  16: core::fmt::rt::Argument::fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/rt.rs:173:76
  17: core::fmt::write
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:1182:21
  18: core::fmt::Formatter::write_fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:1654:13
  19: <ykrt::compile::jitc_yk::aot_ir::Module as core::fmt::Display>::fmt
             at /vol/extra/ltratt/yk/ykrt/src/compile/jitc_yk/aot_ir.rs:269:13
  20: <&T as core::fmt::Display>::fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:2368:62
  21: core::fmt::rt::Argument::fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/rt.rs:173:76
  22: core::fmt::write
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:1182:21
  23: <&mut W as core::fmt::Write::write_fmt::SpecWriteFmt>::spec_write_fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:230:21
  24: core::fmt::Write::write_fmt
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/fmt/mod.rs:235:9
  25: alloc::fmt::format::format_inner
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/alloc/src/fmt.rs:638:14
  26: alloc::fmt::format::{{closure}}
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/alloc/src/fmt.rs:643:34
  27: core::option::Option<T>::map_or_else
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/option.rs:1216:21
  28: alloc::fmt::format
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/alloc/src/fmt.rs:643:5
  29: <ykrt::compile::jitc_yk::JITCYk as ykrt::compile::Compiler>::compile
             at /vol/extra/ltratt/yk/ykrt/src/compile/jitc_yk/mod.rs:94:21
  30: ykrt::mt::MT::queue_compile_job::{{closure}}
             at /vol/extra/ltratt/yk/ykrt/src/mt.rs:592:19
  31: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/core/src/ops/function.rs:250:5
  32: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/8bfcae730a5db2438bbda72796175bba21427be1/library/alloc/src/boxed.rs:2149:9
  33: lock_api::mutex::MutexGuard<R,T>::unlocked
             at /home/ltratt/.cargo/registry/src/index.crates.io-6f17d22bba15001f/lock_api-0.4.12/src/mutex.rs:581:9
  34: ykrt::mt::MT::queue_job::{{closure}}
             at /vol/extra/ltratt/yk/ykrt/src/mt.rs:215:29
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

**Comment by ltratt** [comment]
Actually I think this might relate to @vext01's PR because the root cause is:

```
Self::Unimplemented { tyidx, .. } => {
    let ty = m.type_(*tyidx);
    if ty != &Ty::Void {
        Some(ty)
    } else {
        None
    }
}
```

**Comment by jacob-hughes** [comment]
I think ykllvm got out of sync. I've pushed a submodule update which should fix this now.

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Comment by jacob-hughes** [comment]
I've updated the tests here (49c9bfe793e13a43fd424995778710ef76e0010d) to accommodate the recent logging changes. If you're happy I'll squash.

**Comment by ltratt** [comment]
Please squash.

**Comment by ltratt** [comment]
@vext01 I've squashed this so we can try a merge.

**Comment by vext01** [comment]
@jacob-hughes there's a test failure.

**Comment by ltratt** [comment]
@jacob-hughes Can you try add `NOOPT_VAL(j)` to the test please? That may well fix this.

**Comment by jacob-hughes** [comment]
Fixed in [3e528be](https://github.com/ykjit/yk/pull/1321/commits/3e528be9c7f4add8acaa841f92b99c5f74aa9e54). Tests should pass CI once that's squashed.

**Comment by ltratt** [comment]
Please squash.

**Comment by jacob-hughes** [comment]
Squashed

**Pull request merged**